include/unifex/any_unique.hpp

```diff
@@ -222,65 +208,85 @@ struct _with_type_erased_tag_invoke<
 };
 
-template <
-    typename Derived,
-    typename CPO,
-    bool NoExcept,
-    typename Sig>
-struct _with_forwarding_tag_invoke;
+template <typename CPO, bool NoExcept = false, typename Sig = typename CPO::type_erased_signature_t>
+struct _dispatch_cpo;
 
-template <
-    typename Derived,
-    typename CPO,
-    bool NoExcept = false,
-    typename Sig = typename CPO::type_erased_signature_t>
-using with_forwarding_tag_invoke =
-    typename _with_forwarding_tag_invoke<Derived, CPO, NoExcept, Sig>::type;
+template <typename CPO, typename Ret, typename... Args>
+struct _dispatch_cpo<CPO, false, Ret(Args...) noexcept>
+  : _dispatch_cpo<CPO, true, Ret(Args...)> {};
 
-template <
-    typename Derived,
-    typename CPO,
-    bool NoExcept,
-    typename Ret,
-    typename... Args>
-struct _with_forwarding_tag_invoke<
-    Derived,
-    CPO,
-    NoExcept,
-    Ret(Args...)> {
+template <typename CPO, bool NoExcept, typename Ret, typename... Args>
+struct _dispatch_cpo<CPO, NoExcept, Ret(Args...)> {
   struct type {
-    friend Ret tag_invoke(
-        _overload::base_cpo_t<CPO> cpo,
-        replace_this_t<Args, Derived>... args) noexcept(NoExcept) {
-      auto& wrapper = extract_this<Args...>{}(args...);
-      auto& wrapped = wrapper.value;
-      return std::move(cpo)(replace_this<Args>::get((Args &&) args, wrapped)...);
+    struct _result {
+      UNIFEX_NO_UNIQUE_ADDRESS _overload::base_cpo_t<CPO> cpo;
+
+      template (typename... Args2)
+        (requires ((convertible_to<Args2, Args> || is_this_v<Args>) &&...))
+      Ret operator()(Args2&&... args) && {
+        auto&& wrapper = extract_this<Args...>{}((Args2&&) args...);
+        auto& wrapped = wrapper.value;
+        return std::move(cpo)(replace_this<Args>::get((Args2 &&) args, wrapped)...);
+      }
+    };
+    _result operator()(_overload::base_cpo_t<CPO> cpo) const {
+      return _result{std::move(cpo)};
     }
   };
 };
 
-template <
-    typename Derived,
-    typename CPO,
-    typename Ret,
-    typename... Args>
-struct _with_forwarding_tag_invoke<
-    Derived,
-    CPO,
-    false,
-    Ret(Args...) noexcept>
-  : _with_forwarding_tag_invoke<Derived, CPO, true, Ret(Args...)> {
+template <typename... CPOs>
+struct _dispatch_fn : _dispatch_cpo<CPOs>::type... {
+  using _dispatch_cpo<CPOs>::type::operator()...;
 };
 
-inline const struct deallocate_cpo {
+struct _deallocate_cpo {
   using type_erased_signature_t = void(this_&&) noexcept;
 
-  template(typename T)
-      (requires tag_invocable<deallocate_cpo, T&&>)
+  template <typename T>
   void operator()(T&& obj) const noexcept {
-    tag_invoke(deallocate_cpo{}, (T &&) obj);
+    if constexpr (tag_invocable<_deallocate_cpo, T>) {
+      tag_invoke(_deallocate_cpo{}, (T &&) obj);
+    } else {
+      delete std::addressof(obj);
+    }
   }
-} deallocate {};
+};
+
+template <typename Concrete, typename Allocator, typename... CPOs>
+struct _concrete_impl {
+  struct type final {
```

Comment:
The purpose of this type is to allow us to hold the allocator along with the object.
However, this also means that we'll be instantiating each vtable entry for each combination of Concrete object type and Allocator, where we really only need to instantiate the 'deallocate' entry for this, and should be able to share the vtable entries for the normal CPOs that dispatch to using the object.

I wonder if we could have this type inherit from a base-class, which depends only on Concrete and defines tag_invoke overloads for CPOs, and then only mix in tag_invoke for the deallocate CPO here.

This would require, however, the assumption that the type-erased `void*` for the concrete type reinterpret_cast-able to both the base class and the derived class.

Reply:
Yes, we can probably be smarter about that, but I'm not too concerned about it right now. I anticipate few people to use allocators when constructing `any_unique` instances, and this code isn't hit after my change.